### PR TITLE
Remove delay from say verb caused by typing indicators

### PR DESCRIPTION
**Pull request by KubeRoot**
## About The Pull Request

https://github.com/BeeStation/BeeStation-Hornet/pull/5169 added typing indicators that show up when you're typing something to say or emote from the T and M hotkeys. Because this wraps the original verbs in an argumentless verb and calls input, it delays the window opening until the server replies to the client's message. In case of high ping, and potentially high server load, this can cause a noticable input delay before the window pops up.

To remedy this, the input popup for typing from the hotkeys was replaced with a custom-made window that can be triggered to open completely clientside without server intervention.

I made a beautiful diagram showing the situation before typing indicators, after typing indicators, and after my PR:

Untitled Diagram

I have also made a video in action, with 100ms of latency enabled on my VM, showing how it behaves:

https://streamable.com/6yfgsh

### The bad

Because I'm doing it all clientside and custom, I can't use the builtin window for inputting text. I had to instead create my own window and handle the textbox and buttons. As a downside of limitations of the Input control, used for the textbox, I cannot have the window simply close when you press enter. To work around this, I have made the textbox close when you *release* enter instead, which should be smooth enough.

Also, because it's custom, there will be differences in how it looks. The most notable differences I can think of being:
- There is an orange dream seeker icon on the top bar on the window
- The text in the input is flush with the top edge, instead of being nicely centered with a margin
- The window can be closed - this could be disabled, but I saw some windows of this sort could be closed, others couldn't, so I went with allowing it since I can handle it correctly
- The window doesn't open in quite the right location if the main window is maximized or anchored (win+right)

### The big explanation

Below is the entire previous description of this PR, before I finalised it and set as ready to review. It contains more explanations on how it works and why things are the way they are. It's quite lengthy and chaotic though, so not sure how much use it is:

<details>

<summary>The big description with full explanation</summary>

Previously, without typing indicators, the verb called when you press `T` took a text message as an argument. This displayed a textbox where you could type the message you want to say. This was handled by BYOND, and would happen before the server knew you were using a verb at all, which means there was no delay due to ping.

The typing indicators, however, wrapped the original code in a wrapper verb that doesn't take an argument and instead calls `input`. This was necessary due to needing to create/remove the typing indicator before/after the input is shown.
However, because the textbox was no longer handled by BYOND directly, this added a delay - your client needs to send a message to the server to invoke the verb, then the server needs to send a message to the client to display the textbox, causing a delay of a full roundtrip based on ping.

After some asking around, I found one way to remove the delay, which is implemented in this PR.

Instead of simply using a verb, the `T` keybind now does two things:
1. Calls a `.start_typing` verb to show a typing indicator
2. Uses `.winset` to show a custom-made textbox, directly on the client, avoiding the delay from doing this from a verb.

There are, of course, caveats:
1. Clientside macros are very chaotic. I'd say a lot of how this works makes no sense, but doing it in reasonable ways didn't work. Maintaining this might be painful.
2. The first time you use the hotkey the window has to be created from the server, for this reason the keybind also calls an additional verb every time you press it. Afterwards, however, the window is hid instead of closing, preventing delay.
3. **The big one.** There is instead a delay before the window closes after you press enter. This is because of how textboxes work in BYOND, which I'll explain further at the end.
  - **EDIT:** This was somewhat fixed. Instead of closing when the server tells the window to close, the window now closes only after you release enter.

### For a quick summary of how this all works:

There is a new window defined in `interface/skin.dmf`, containing a generic window with an Input and two Buttons. A new proc on client called `create_input_window` can be used to create a window with a specific ID, title and verbs to call on accept/cancel. This proc creates a copy of the generic window, setting the right title and commands on its controls, then it displays the window.

The three controls on the window have commands set on them - the buttons use `.winset` to both call a proc and hide the window.
The Input, however, is where it gets tricky.
Inputs in BYOND have two modes: command and no-command.
- When in command mode, the Input does input validation, making sure you're inputting a correct verb and running autocompletion when you press space. When you press enter in this mode, it runs the text you typed in as a command, prepending whatever was set as its `command` field.
- When in no-command mode, you can type anything you want into the Input, and space does not have any special behavior. However, the enter button doesn't do anything, and macro sets seem to not work.

For this, the Input is set to command mode, with the `command` field set to `.say \"`.

Finally, the `T` hotkey was changed to this monstrosity:

```
"T" = "\".winset \\\"command=.start_typing;command=.init_say;saywindow.is-visible=true;saywindow.input.focus=true;saywindow.input.text=\\\"\\\"\\\"\"",
```

Which does the following things:
- Invoke `.start_typing` to display the typing indicator
- Invoke `.init_say` to create a Say window if there isn't one already
- set `saywindow.is-visible=true` to display the window if it exists
- set `saywindow.input.focus=true` to focus on the input textbox
- set `saywindow.input.text=\"\"` to clear the input textbox

### The caveat

Now, the big issue: there is a delay before the window closes when you press enter. This is because of how Inputs work - we can't have it accept input when you press enter when in no-command mode, so we have to use command mode. However, while in command mode, we can't have it run convoluted `.winset` commands while being able to type in spaces. This is because for the Input to accept space as a valid character, you have to either be at the end of a valid verb or argument, or inside a string.
Because of this, the command has to be simple, in this case `.say \"` - which means the Input knows it is in the middle of a string argument, allowing you to type spaces, but prevents us from running clientside macros as a result of pressing enter.

What this means is, while the window can open immediately, if submitted by pressing enter, it has to be closed by the server instead.

Because of this, the scourge of the typing indicator input delay isn't fully removed, instead being moved to closing instead of opening.

It **might be** possible to remove this delay, but I haven't managed in a couple hours of trying.

### Partial solution

I have figured out a way to bypass the issue of closing input lag. While you cannot run a macro on pressing enter when you have a message typed in, you can run a macro when the input is empty. And while the input obviously isn't empty when there's something in the textbox, as soon as you press enter, the textbox is cleared. This means... you can run a macro when you *release* enter, after the textbox clears, that will hide the window.

Along with that, I have added macros that cancel when you press escape, or press enter with an empty textbox, enabled the close button that also cancels. I have added centering when the window initially opens, putting it in the middle of the main game window... which only works correctly if your window is not anchored to the side (win+right, for example), because BYOND.

I have also modified the `erase_all_macros` proc on the client, used by the `input` subsystem when resetting all client macrosets - it now ignores macrosets with the name starting with `persist_`. This could certainly be done better, but for now it should hopefully suffice. The reason for this change is that the macro set for the say window could 

### Videos

I recorded a couple videos of it in action. There's nothing particularly interesting to see, but the second video more importantly has 100ms of latency added showing the issue with closing the window with enter.

https://streamable.com/03i7tz
~~https://streamable.com/oe1hs5~~

Here's a new video with latency: https://streamable.com/6yfgsh

</details>

## Why It's Good For The Game

Input delay is bad

## Changelog
:cl:
tweak: The "Say" window no longer has input lag when opening it with the hotkey.
tweak: The "Say" window now has input lag when closing it with enter.
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
HRRRRRRG BASED. 

Thanks for committing the time to fix this.
image

**Comment by francinum** [comment]
I am completely unqualified to review this PR, ike.

**Comment by KubeRoot** [comment]
First thing that showed up on testing... Seems like the interface scales badly with dpi scaling. Most interfaces in the game use browser controls to display the actual content, which presumably allows them to scale properly. Because I use DPI scaling myself, and matched the window on my machine, a lot of users without DPI scaling are seeing the window and controls as significantly bigger than it should be.

This shouldn't impact gameplay, but it is a notable annoyance. It should also be possible to work around partially by using a browser, which means the window itself would still be bigger than it should be, but the controls wouldn't look mismatched.

I'm unsure if this is a BYOND bug that could get fixed, or an intentional design choice, since ShareX tells me the custom window has the same size regardless of DPI scaling and it's the builtin input window that changes size.

EDIT: There's also currently an issue - if you close the window using the X button, it shows a `.stop_typing "text"` window because I forgot to add quotes to the command when adding the argument.

**Comment by AnCopper** [comment]
image
You made the chat bar big, bulky and ugly.

**Comment by KubeRoot** [comment]
> You made the chat bar big, bulky and ugly.

Correction: BYOND did. It looks like this for me:
image

More specifically, BYOND does DPI scaling badly, so for me at 125% DPI scaling it looks fine

**Comment by AnCopper** [comment]
There also seems to be a tiny delay now between sending the message and the chat bar disappearing.

**Comment by KubeRoot** [comment]
> There also seems to be a tiny delay now between sending the message and the chat bar disappearing.

Yes, the PR description explains it in more detail, but basically the best I can do is make the window close when you release enter after sending a message.

**Comment by KubeRoot** [comment]
Figured out a way to detect DPI scaling - I can compare the window's inner and outer size to determine the size of the titlebar, and deduce the DPI scaling used from that. This means the window should have a size consistent with BYOND's own textbox at different DPI scaling.

TL;DR - No more huge windows!

**Comment by brrtshka42** [comment]
seems to cause massive lag spikes every time I try to move. never had abnormal lag on bee before, it only started happening after the testmerge, doesn't happen to me on any other servers, and [some other people on the forums](https://forums.beestation13.com/t/ping-related-issues/14824) have mentioned having very similar problems. i've tried using multiple different installs of BYOND, as well. i don't really understand why it's happening but it's pretty much rendered the game unplayable for me. i thought i might as well post it here.

**Comment by KubeRoot** [comment]
> seems to cause massive lag spikes every time I try to move. never had abnormal lag on bee before, it only started happening after the testmerge, doesn't happen to me on any other servers, and [some other people on the forums](https://forums.beestation13.com/t/ping-related-issues/14824) have mentioned having very similar problems. i've tried using multiple different installs of BYOND, as well. i don't really understand why it's happening but it's pretty much rendered the game unplayable for me. i thought i might as well post it here.

It seems Acacia doesn't have the testmerge running, can you try playing there and see if you have the same issues? I'll also ping somebody higher up to mention this, though I don't see how this could be causing it... But then again, it's BYOND.

**Comment by francinum** [comment]
The recent issues related to performance seem to be related to unlucky routing as VPNs seem to help the situation.

**Comment by ike709** [comment]
YOLO
